README.md
# AWS Lambda Scala Example Project

[ ![Build Status] [travis-image] ] [travis]  [ ![Release] [release-image] ] [releases] [ ![License] [license-image] ] [license]

## Introduction

This is an example [AWS Lambda] [aws-lambda] Scala application for processing a [Kinesis] [aws-kinesis] stream of events ([introductory blog post] [blog-post]). It reads the stream of simple JSON events generated by our event generator. Our AWS Lambda function aggregates and buckets events and stores them in [DynamoDB] [aws-dynamodb].

This was built by the Data Science team at [Snowplow Analytics] [snowplow], who use AWS Lambda in their projects.

**Running this requires an Amazon AWS account, and will incur charges.**

_See also:_ [AWS Lambda Node.js Project][aws-lambda-nodejs-example-project] | [Spark Streaming Example Project][spark-streaming-example-project]

## Overview

We have implemented a super-simple analytics-on-write stream processing job using AWS Lambda. Our AWS Lambda function, written in Scala that runs on the Java8 JVM, reads a Kinesis stream containing events in a JSON format:

```json
{
  "timestamp": "2015-06-05T12:54:43.064528",
  "eventType": "Green",
  "id": "4ec80fb1-0963-4e35-8f54-ce760499d974"
}
```

Our job counts the events by `eventType` and aggregates these counts into 1 minute buckets. The job then takes these aggregates and saves them into a table in DynamoDB:

![dynamodb-table-image][dynamodb-table-image]

## Developer Quickstart

Assuming git, [Vagrant] [vagrant-install] and [VirtualBox] [virtualbox-install] installed:

```bash
 host$ git clone https://github.com/snowplow/aws-lambda-scala-example-project.git
 host$ cd aws-lambda-scala-example-project
 host$ vagrant up && vagrant ssh
guest$ cd /vagrant
guest# sbt assembly
```

## Tutorial

You can follow along in [the release blog post] [blog-post] to get the project up and running yourself.

The following steps assume that you are running inside Vagrant, as per the Developer Quickstart above.

### 1. Setting up AWS credentials

First we need to configure a default AWS profile:

```bash
$ aws configure
AWS Access Key ID [None]: ...
AWS Secret Access Key [None]: ...
Default region name [None]: us-east-1
Default output format [None]: json
```

### 2. Setup Amazon Kinesis, DynamoDB, and IAM Role

Now we create our Kinesis event stream:

```bash
$ inv create_kinesis_stream my-stream
Kinesis Stream [my-stream] not active yet
Kinesis Stream [my-stream] not active yet
Kinesis Stream [my-stream] not active yet
Kinesis successfully created.
```

Now create our DynamoDB table:

```bash
$ inv create_dynamodb_table default us-east-1 my-table
```

Now we can create our IAM role. We will be using [CloudFormation] [cloudformation] to make our new role. Using `inv create_role`, we can create it like so:

```bash
$ inv create_role
arn:aws:cloudformation:us-east-1:84412349716:stack/LambdaStack/23a341eb0-4162-11e5-9d4f-0150b34c7c
Creating roles
Still creating
Giving Lambda proper permissions
Trying...
Created role
```

### 3.  Build the Scala project jar

Let's build our Scala project into a fully self contained jar file.

```bash
$ sbt assembly
info] Loading project definition from /aws-lambda-scala-example-project/project
[info] Set current project to aws-lambda-scala-example-project (in build file:/aws-lambda-scala-example-project/)
[info] Including from cache: scala-reflect-2.11.4.jar
...
[warn] Merging 'rootdoc.txt' with strategy 'first'
[warn] Strategy 'discard' was applied to 62 files
[warn] Strategy 'first' was applied to a file
[info] SHA-1: 96401bbad71968267ccea4c479a7d39093ef8988
[info] Packaging /Volumes/DataDrive/dev/aws-lambda-scala-example-project/target/scala-2.11/aws-lambda-scala-example-project-0.2.0.jar ...
[info] Done packaging.
[success] Total time: 59 s, completed 13-Aug-2015 10:40:05 AM
```

### 4.  Upload project jar to Amazon S3.

We will create a S3 Bucket for the jar file to be picked up by AWS Lambda.  We will upload the jar file to the Amazon S3 service using our custom uploader `inv upload_s3`.

```bash
$ inv upload_s3
Jar uploaded to S3 aws_scala_lambda_bucket
```

### 5.  Configure AWS Lambda service

Now that we have built the project, and uploaded the jar file to the AWS Lambda service, we need to configure the Lambda service to watch for event traffic from our AWS Kinesis stream named `my-stream`.

```bash
$ inv create_lambda
Creating AWS Lambda function.
{
    "FunctionName": "ProcessingKinesisLambdaDynamoDB",
    "CodeSize": 38042279,
    "MemorySize": 1024,
    "FunctionArn": "arn:aws:lambda:us-east-1:842349429716:function:ProcessingKinesisLambdaDynamoDB",
    "Handler": "com.snowplowanalytics.awslambda.LambdaFunction::recordHandler",
    "Role": "arn:aws:iam::842340234716:role/LambdaStack-LambdaExecRole-7G57P4M2VV5P",
    "Timeout": 60,
    "LastModified": "2015-08-13T19:39:46.730+0000",
    "Runtime": "java8",
    "Description": ""
}
```

Now we can associate our Lambda with our Kinesis stream:

```bash
$ inv configure_lambda my-stream
Configured AWS Lambda service.
Added Kinesis as event source for Lambda function.
```

### 6. Sending events to Kinesis

We need to start sending events to our new Kinesis stream. We have created a helper method to do this - run the below and leave it running:

```bash
$ inv generate_events default us-east-1 my-stream
Event sent to Kinesis: {"timestamp": "2015-06-05T12:54:43.064528", "type": "Green", "id": "4ec80fb1-0963-4e35-8f54-ce760499d974"}
Event sent to Kinesis: {"timestamp": "2015-06-05T12:54:43.757797", "type": "Red", "id": "eb84b0d1-f793-4213-8a65-2fb09eab8c5c"}
Event sent to Kinesis: {"timestamp": "2015-06-05T12:54:44.295972", "type": "Yellow", "id": "4654bdc8-86d4-44a3-9920-fee7939e2582"}
...
```

### 7. Monitoring your job

First head over to the AWS Lambda service console, then review the logs in CloudWatch.

Finally, let's check the data in our DynamoDB table. Make sure you are in the correct AWS region, then click on `my-table` and hit the `Explore Table` button:

![dynamodb-table-image][dynamodb-table-image]

For each **BucketStart** and **EventType** pair, we see a **Count**, plus some **CreatedAt** and **UpdatedAt** metadata for debugging purposes. Our bucket size is 1 minute, and we have 5 discrete event types, hence the matrix of rows that we see.

## Roadmap

* Expanding our analytics-on-write thinking into our new [Icebucket] [icebucket] project

## Credits

* Ian Meyers and his [Amazon Kinesis Aggregators][amazon-kinesis-aggregators] project, a true inspiration for streaming analytics-on-write

## Copyright and license

AWS Lambda Scala Example Project is copyright 2015 Snowplow Analytics Ltd.

Licensed under the [Apache License, Version 2.0] [license] (the "License");
you may not use this software except in compliance with the License.

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

[travis]: https://travis-ci.org/snowplow/aws-lambda-scala-example-project
[travis-image]: https://travis-ci.org/snowplow/aws-lambda-scala-example-project.png?branch=master
[license-image]: http://img.shields.io/badge/license-Apache--2-blue.svg?style=flat
[license]: http://www.apache.org/licenses/LICENSE-2.0
[release-image]: http://img.shields.io/badge/release-0.2.0-blue.svg?style=flat
[releases]: https://github.com/snowplow/aws-lambda-scala-example-project/releases
[grunt-image]: https://cdn.gruntjs.com/builtwith.png

[spark-streaming-example-project]: https://github.com/snowplow/spark-streaming-example-project
[aws-lambda-nodejs-example-project]: https://github.com/snowplow/aws-lambda-nodejs-example-project
[vagrant-install]: http://docs.vagrantup.com/v2/installation/index.html
[virtualbox-install]: https://www.virtualbox.org/wiki/Downloads

[blog-post]: http://snowplowanalytics.com/blog/2015/08/20/aws-lambda-scala-example-project-0.1.0-released/
[dynamodb-table-image]: /docs/dynamodb-table-image.png?raw=true

[aws-cloudformation]: http://aws.amazon.com/cloudformation
[aws-lambda]: http://aws.amazon.com/lambda/
[aws-kinesis]: http://aws.amazon.com/kinesis/
[aws-dynamodb]: http://aws.amazon.com/dynamodb
[amazon-kinesis-aggregators]: https://github.com/awslabs/amazon-kinesis-aggregators

[snowplow]: http://snowplowanalytics.com
[icebucket]: https://github.com/snowplow/icebucket


CHANGELOG
Version 0.2.0 (2015-01-25)
--------------------------
Specify SBT version in build.properties (#8)
Replace .sbt build with Scala build project (#9)
Set build artifact file extension to '.jar' (#7)
Fix travis builds, update to jdk8 (#5)

Version 0.1.0 (2015-08-20)
--------------------------
Initial release


tasks.py
# Copyright (c) 2015 Snowplow Analytics Ltd. All rights reserved.
#
# This program is licensed to you under the Apache License Version 2.0,
# and you may not use this file except in compliance with the Apache License Version 2.0.
# You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the Apache License Version 2.0 is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

import datetime, json, uuid, time
from functools import partial
from random import choice
from invoke import run, task
import boto
from boto import kinesis
import boto.dynamodb2
from boto.dynamodb2.fields import HashKey, RangeKey, KeysOnlyIndex, GlobalAllIndex
from boto.dynamodb2.table import Table
from boto.dynamodb2.types import NUMBER
import boto.cloudformation
import time
import math, os
from filechunkio import FileChunkIO

# setting for AWS Lambda and Lambda-Exec-Role
REGION = "us-east-1"
IAM_ROLE_ARN = ""
IAM_ROLE = ""
POLICY = """{
    "Statement":[{
    "Effect":"Allow",
    "Action":["*"],
    "Resource":["*"]}]}"""
POLICY_NAME = "AdministratorAccess"
STACK_NAME = "LambdaStack"
TEMPLATE_URL = "https://snowplow-hosted-assets.s3.amazonaws.com/third-party/aws-lambda/lambda-admin.template"
CAPABILITIES = ["CAPABILITY_IAM"]
JARFILE = "./target/scala-2.11/aws-lambda-scala-example-project-0.2.0.jar"
S3_BUCKET = "aws_scala_lambda_bucket"
S3_KEY = os.path.basename(JARFILE)
FUNCTION_NAME = "ProcessingKinesisLambdaDynamoDB"
# Selection of EventType values
COLORS = ['Red','Orange','Yellow','Green','Blue']
# DynamoDB settings
THROUGHPUT_READ = 20
THROUGHPUT_WRITE = 20


# AWS Kinesis Data Generator
def picker(seq):
    """
    Returns a new function that can be called without arguments
    to select and return a random color
    """
    return partial(choice, seq)

def create_event():
    """
    Returns a choice of color and builds and event
    """
    event_id = str(uuid.uuid4())
    color_choice = picker(COLORS)

    return (event_id, {
      "id": event_id,
      "timestamp": datetime.datetime.now().isoformat(),
      "eventType": color_choice()
    })

def write_event(conn, stream_name):
    """
    Returns the event and event event_payload
    """
    event_id, event_payload = create_event()
    event_json = json.dumps(event_payload)
    conn.put_record(stream_name, event_json, event_id)
    return event_json

@task
def upload_s3():
    """
    Upload jar file to s3
    """
    source_path = JARFILE
    source_size = os.stat(source_path).st_size

    # create bucket
    import boto
    conn = boto.connect_s3()
    bucket = conn.create_bucket(S3_BUCKET)

    # upload
    c = boto.connect_s3()
    b = c.get_bucket(S3_BUCKET)
    # Create a multipart upload request
    mp = b.initiate_multipart_upload(os.path.basename(source_path))

    # Use a chunk size of 5 MiB
    chunk_size = 5242880
    chunk_count = int(math.ceil(source_size / float(chunk_size)))

    # Send the file parts, using FileChunkIO to create a file-like object
    # that points to a certain byte range within the original file. We
    # set bytes to never exceed the original file size.
    for i in range(chunk_count):
        offset = chunk_size * i
        bytes = min(chunk_size, source_size - offset)
        with FileChunkIO(source_path, 'r', offset=offset,
                         bytes=bytes) as fp:
            mp.upload_part_from_file(fp, part_num=i + 1)

    # Finish the upload
    mp.complete_upload()
    print("Jar uploaded to S3 bucket " + S3_BUCKET)

@task
def create_role():
    """
    Creates IAM role using CloudFormation for AWS Lambda service
    """
    client_cf = boto.cloudformation.connect_to_region(REGION)
    response = client_cf.create_stack(
        stack_name=STACK_NAME,
        template_url=TEMPLATE_URL,
        capabilities=CAPABILITIES
    )
    print response
    time.sleep(7)
    print "Creating roles"
    time.sleep(7)
    print "Still creating"
    time.sleep(7)
    print "Giving Lambda proper permissions"
    # get name of LambdaExecRole
    client_iam = boto.connect_iam()
    roles = client_iam.list_roles()
    list_roles = roles['list_roles_response']['list_roles_result']['roles']
    for i in range(len(list_roles)):
        if STACK_NAME+"-LambdaExecRole" in list_roles[i].arn:
            IAM_ROLE = list_roles[i].role_name
    print "Trying..."
    # grants Admin access to LambdaExecRole to access Cloudwatch, DynamoDB, Kinesis
    client_iam.put_role_policy(IAM_ROLE, POLICY_NAME, POLICY)
    print "Created role"


@task
def generate_events(profile, region, stream):
    """
    load demo data with python generator script for SimpleEvents
    """
    conn = kinesis.connect_to_region(region, profile_name=profile)
    while True:
        event_json = write_event(conn, stream)
        print "Event sent to Kinesis: {}".format(event_json)

@task
def create_lambda():
    """
    Create aws-lambda-scala-example-project AWS Lambda service
    """
    # TODO: switch to use all boto
    IAM_ROLE_ARN = get_iam_role_arn()
    print("Creating AWS Lambda function.")
    run("aws lambda create-function --region {} \
                                    --function-name {} \
                                    --code S3Bucket={},S3Key={} \
                                    --role {} \
                                    --handler com.snowplowanalytics.awslambda.LambdaFunction::recordHandler \
                                    --runtime java8 --timeout 60 --memory-size 1024".format(REGION, FUNCTION_NAME, S3_BUCKET, S3_KEY, IAM_ROLE_ARN), pty=True)

def get_iam_role_arn():
    client_iam = boto.connect_iam()
    roles = client_iam.list_roles()
    list_roles = roles['list_roles_response']['list_roles_result']['roles']
    for i in range(len(list_roles)):
        if STACK_NAME+"-LambdaExecRole" in list_roles[i].arn:
            IAM_ROLE_ARN = list_roles[i].arn
    return IAM_ROLE_ARN

@task
def configure_lambda(stream):
    """
    Configure Lambda function to use Kinesis
    """
    print("Configured AWS Lambda service")
    IAM_ROLE_ARN = get_iam_role_arn()
    aws_lambda = boto.connect_awslambda()
    event_source = kinesis_stream(stream)
    response_add_event_source = aws_lambda.add_event_source(event_source,
                                                            FUNCTION_NAME,
                                                            IAM_ROLE_ARN,
                                                            batch_size=100,
                                                            parameters=None)
    event_source_id = response_add_event_source['UUID']

    while response_add_event_source['IsActive'] != 'true':
        print('Waiting for the event source to become active')
        sleep(5)
        response_add_event_source = aws_lambda.get_event_source(event_source_id)

    print('Added Kinesis as event source for Lambda function')


@task
def create_dynamodb_table(profile, region, table):
    """
    DynamoDB table creation with AWS Boto library in Python
    """
    connection = boto.dynamodb2.connect_to_region(region, profile_name=profile)
    aggregate = Table.create(table,
                             schema=[
                                 HashKey("BucketStart"),
                                 RangeKey("EventType"),
                             ],
                             throughput={
                                 'read': THROUGHPUT_READ,
                                 'write': THROUGHPUT_WRITE
                             },
                             connection=connection
                             )

@task
def create_kinesis_stream(stream):
    """
    Creates our Kinesis stream
    """
    kinesis = boto.connect_kinesis()
    response = kinesis.create_stream(stream, 1)
    pause_until_kinesis_active(stream)
    print("Kinesis successfully created")

def pause_until_kinesis_active(stream):
    kinesis = boto.connect_kinesis()
    # Wait for Kinesis stream to be active
    while kinesis.describe_stream(stream)['StreamDescription']['StreamStatus'] != 'ACTIVE':
        print('Kinesis stream [' + stream + '] not active yet')
        time.sleep(5)

def kinesis_stream(stream):
    """
    Returns Kinesis stream arn
    """
    kinesis = boto.connect_kinesis()
    return kinesis.describe_stream(stream)['StreamDescription']['StreamARN']

@task
def describe_kinesis_stream(stream):
    """
    Prints status Kinesis stream
    """
    print("Created: ")
    print(kinesis_stream(stream))
